FLEDGE.md

```diff
@@ -281,6 +288,42 @@ Once all of a component auction's bids have been scored by the component auction
 The ultimate winner of the top-level auction is the single bid the top-level seller script gives the highest score. This may either be the winning bid of one of the component auctions, or a bid from one of the `interestGroupBuyers` in the `AuctionConfig` of the top-level auction. Those bids will be scored directly by the top-level seller script without having to win any component auction.
 
 
+#### 2.5 extraSignals Hidden from the Page (for Auction Worklets Only)
+
+Auction config parameter values are visible to any scripts, first or third party, that are running on the frame that called `runAdAuction()`. For instance, a script could assign a new proxy function to `navigator.runAdAuction()`.
+
+For auction fairness, sellers might not want to reveal seller and per-buyer signals to other parties.
+
+
+The optional `extraSellers` field can be used to pass signals to the auction, similar to `sellerSignals` and `perBuyerSignals`. The difference is that `extraSignals` content cannot be viewed by scripts running on the page -- the signals may only be read by the respective intended auction participants. If present, `extraSellers` should be an https URL of a resource in a [subresource bundle](https://github.com/WICG/webpackage/blob/main/explainers/subresource-loading.md) that has been loaded by the current document, whose contents are of type `application/json`, with the following response headers: `X-Allow-Fledge: true` and `X-FLEDGE-Auction-Only: true`.
+
+The JSON response should be a dictionary, whose keys are the origins of the auction participants (including both the seller and individual buyers), and whose values are the signals to be delivered to each respective auction participant.
+
+```
+{
+  'seller.com': {...},
```

Comment:
This scheme requires we either do JSON handling in the browser process, or pass the entire JSON dictionary (complete with data intended for other origins) to the untrusted origin-scoped processes that run worklets. That seems problematic?

Reply:
Yes, it's a trade-off.

We could specify a dictionary for each extra signal separately in the auction config (similar to perBuyerSignals, but with bundle URLs as the dictionary values, instead of objects as values), but the publisher (origin that calls runAdAuction()) controls the auction config, and it isn't trusted to send the signals to the correct party. Perhaps this is OK? The original #119 issue was about preventing the page from reading the signals -- if signals are sent to the wrong worklet, it shouldn't be possible for the worklet to then send those signals back to the page.

We could also introduce a special header that specifies the intended recipient of each signal, but internal reviewers thought it was too complex to introduce such header semantics.

Of course, the downside of the current approach is that the browser process has to keep the JSON in memory, process it, and send portions to the worklets.

Reply:
Just brainstorming some other options here:

We could have one dict be an index with links to other dicts (and handle passing around the URLs in the browser process).

We could use a fixed url prefix (e.g., the seller gives us "https://seller.com/prefix", and then we append "?buyer=https://buyer.com" or "seller=..." for seller (or just origin=, and not distinguish between the two).

Hrm...I'm not seeing a huge number of choices, with throwing the seller worklet at it, which has perf implications.

Reply:
> We could have one dict be an index with links to other dicts (and handle passing around the URLs in the browser process).

Perhaps, but this seems a bit complicated -- we're now doing network requests in 2 places, not just one. It would reduce the amount of copying.

> We could use a fixed url prefix (e.g., the seller gives us "https://seller.com/prefix", and then we append "?buyer=https://buyer.com/" or "seller=..." for seller (or just origin=, and not distinguish between the two).We could use a fixed url prefix (e.g., the seller gives us "https://seller.com/prefix", and then we append "?buyer=https://buyer.com/" or "seller=..." for seller (or just origin=, and not distinguish between the two).

If the untrusted worklet process knew the URL of some other signal somehow, couldn't it fetch that? (I think this would require a compromised worklet). A proxy could be created in the document to retrieve the URLs of all the signals, so they wouldn't be secret.

Also, in the Chromium implementation, for any approach where the worklet does the fetching, somehow we'd need to authorize the worklets to receive X-FLEDGE-Auction-Only responses, since currently we only allow trusted URLLoaderFactory's to receive X-FLEDGE-Auction-Only responses, whereas the worklet is inherently untrusted.

Reply:
> > We could use a fixed url prefix (e.g., the seller gives us "https://seller.com/prefix", and then we append "?buyer=https://buyer.com/" or "seller=..." for seller (or just origin=, and not distinguish between the two).We could use a fixed url prefix (e.g., the seller gives us "https://seller.com/prefix", and then we append "?buyer=https://buyer.com/" or "seller=..." for seller (or just origin=, and not distinguish between the two).
> 
> If the untrusted worklet process knew the URL of some other signal somehow, couldn't it fetch that? (I think this would require a compromised worklet). A proxy could be created in the document to retrieve the URLs of all the signals, so they wouldn't be secret.

No.  That's the entire point of the proxy class:  https://source.chromium.org/chromium/chromium/src/+/main:content/browser/interest_group/auction_url_loader_factory_proxy.h

> Also, in the Chromium implementation, for any approach where the worklet does the fetching, somehow we'd need to authorize the worklets to receive X-FLEDGE-Auction-Only responses, since currently we only allow trusted URLLoaderFactory's to receive X-FLEDGE-Auction-Only responses, whereas the worklet is inherently untrusted.

Also the job of the proxy class.  It already uses trusted URLLoaderFactories for most requests.

Reply:
Ah, right, I was forgetting that we verify and proxy the requests.

So, maybe the prefix idea is the approach we want to take? Trying to think if there are any downsides beyond just needing to rewrite the existing code...

The proxy would be trusted to receive X-FLEDGE-Auction-Only, and it would ensure that every worklet gets only the signals destined for it, by following the prefix scheme.

We'd do more total network requests, but these are all going to the in-network service bundle, and we'd save the copy to worklet Mojo overhead.

> Hrm...I'm not seeing a huge number of choices, with throwing the seller worklet at it, which has perf implications.

Not sure what "with throwing the seller worklet at it" means? Having the seller worklet do all the parsing, even for the buyers?

Reply:
I don't think it particularly matters - I'm fine with any option, here.  If we only serve these from a web bundle, we could even just append "&foo" if there's a "?", and "?foo" if there's not.  This isn't something I'd be comfortable doing for HTTPS requests, though, since a publisher could potentially provide an incorrect prefix (e.g. "https://ssp.com/?perBuyerSignals=foo", and then if we append "&perBuyerSignals=foo", the wrong signals could be sent.

Hrm...  Are we concerned about the publisher deliberately using the wrong web bundle, if different web bundles are used for different pages on the same publisher?

e.g., publisher has articles A and B which the SSP want to send different signals on, but the publisher loads article B's web bundle for article A.  A scheme that avoids that seems difficult, though I guess the SSP will get to see the URL prefix in auctionConfig, and could theoretically validate it there.

Reply:
Actually, maybe not?  I don't think we provide the full URL to scripts via browserSignals.

Reply:
>  I guess the SSP will get to see the URL prefix in auctionConfig, and could theoretically validate it there.

Yes -- we'd need to make sure to pass that to the worklet. If validation fails, the seller worklet could fail the auction. But, I don't think the browser signals includes the top-level URL path, so I'm not sure how practical this is?

Reply:
I've pushed a new commit that addresses the original issue of this comment, PTAL.

Reply:
The latest push seems reasonable to me, at least.